### ISSUE TITLE
tox --parallel can "time out" on CI providers

### ISSUE THREAD

**Issue by schinckel** [tox-dev/tox#1142]
Firstly, thank for tox, I love it.

I've been using detox for some time to run tests in parallel, and was very excited to see the capability added in to tox itself. However, there is one place where the detox behaviour was nicer.

We have a bunch of tests that take about half an hour to run in one test environment, and another process that takes a similar amount of time to run in a different environment (that tests a part of our build process).

Codeship, the CI provider we use will kill a process that has not produced any output in the last 10 minutes, and determine this is a failure.

`tox --parallel auto` can trigger this, so we turned on `--parallel-live`, which outputs the data from the various test cases, but (as expected) the data is interleaved.

`detox` used to display the environment names, and an ASCII spinner while the tests were running, to show that output was being processed. Then, if any tests in an environment failed, it output the test results for that environment, otherwise it was a (much shorter) message about success.

This is problematic because having all of the data from all test runs is currently required to stop the timeout, but also makes it really hard to see which test cases failed (and could possibly result in interleaved tracebacks in the case of failures happening at the same time in different environments).

**Comment by gaborbernat** [comment]
tox has the same functionality, we'll print the spinner and outcome of test results. Why is the spinner ignored?

**Comment by obestwalter** [comment]
Maybe unicode woes? The old spinner was ASCII.

**Comment by schinckel** [comment]
Yes, running tests locally I can see it is a much prettier unicode spinner.

When I watch the tests run on codeship (without the `--parallel-live` setting), then indeed it doesn't show the spinner.

I'll chase it up with them.

**Comment by gaborbernat** [comment]
could be the spinner gets ignored, which makes it seem static, I'll look into a way to enable some ASCI fallback🤔

**Comment by schinckel** [comment]
I've got a fork I'm trying that with: I just have a spinner that uses `+` and `x` for now ;)

Once I get that build onto my CI server, I'll report back.

**Comment by schinckel** [comment]
Okay, it looks like it's got something to do with the mechanism being used to update the spinner.

I wonder how the detox spinner was updating the value, compared to how the new spinner is implemented.

A support person from Codeship said:

> yeah unfortunately the way our system captures logs from the build machines doesn't really work with the fancy output from some tools. When tools are continually updating the same exact line, rather than printing a new one, it is especially problematic.

But, again, it used to work in detox.

**Comment by gaborbernat** [comment]
create a plugin that prints a new line every 10 second?😅

**Comment by schinckel** [comment]
FWIW, bitbucket pipelines handles it in an interestingly different way: https://bitbucket.org/schinckel/django-keyset-pagination/addon/pipelines/home#!/results/23

**Comment by themr0c** [comment]
Hi, on bamboo the new UTF-8 spinner is also causing trouble. It is producing lots of garbage output: each move of the spinner is added to the log. 

It would be very nice to have a way to disable the spinner.

**Comment by gaborbernat** [comment]
We need to add an environment variable that disables it, which one could set for some CIs.

**Comment by gaborbernat** [comment]
I think this should now be fixed with version 4 (rewrite branch). If still reproducible please post it and we'll reopen it.

**Comment by robert-schmidtke** [comment]
> I think this should now be fixed with version 4 (rewrite branch). If still reproducible please post it and we'll reopen it.

Hi @gaborbernat. First of all I was happy to see the first beta for version 4 out there!

I have a bunch of tests I run using `--parallel` in my CI pipeline (Bitbucket). There are a total of 6 environments that should execute in parallel, differing only in the version of one dependency (`awswrangler209` thru `awswrangler214`).

Both with the latest tox 3 and 4b1, the build sometimes times out. Looking at the logs, the builds that time out only produce logs for 5 out of those 6 environments, and never show the final summary.

Successful log:
```
✔ OK awswrangler210 ...
...
✔ OK awswrangler211 ...
...
✔ OK awswrangler213 ...
...
✔ OK awswrangler209 ...
...
✔ OK awswrangler212 ...
...
✔ OK awswrangler214 ...
...
___________________________________ summary ____________________________________
  awswrangler209: commands succeeded
  awswrangler210: commands succeeded
  awswrangler211: commands succeeded
  awswrangler212: commands succeeded
  awswrangler213: commands succeeded
  awswrangler214: commands succeeded
  congratulations :)
```

Instead I sometimes only see this (note only 5 enviroments are shown, with the summary missing):
```
✔ OK awswrangler210 ...
...
✔ OK awswrangler211 ...
...
✔ OK awswrangler213 ...
...
✔ OK awswrangler209 ...
...
✔ OK awswrangler214 ...
...
```

After that the build hangs and never gets to finish the missing environment / showing the final log.

Is this related to this issue here, or should I open a new ticket? If so, what other information would you need from me to diagnose this better?

Thanks a lot!

**Comment by robert-schmidtke** [comment]
Turns out this was actually not related to `tox`, but rather me using [Python signal](https://docs.python.org/3/library/signal.html) in my tests that would sometimes not fire in the right order and cause tests to hang. This would also happen when running tests outside of `tox` and CI, albeit not as often. It was just that it became most obvious when running in parallel and on CI.

I removed the tests relying on `signal` (I used to force a timeout by sending `SIGALRM`) and the issues vanished.

Apologies for polluting this issue.
